aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -0,0 +1,25 @@
+#include "TensorTransformations.h"
+
+#include "ATen/NativeFunctions.h"
+
+namespace at {
+namespace native {
+
+Tensor flip_cpu(const Tensor& self, IntList dims) {
+  const int64_t total_dims = self.dim(), flip_dims_size = dims.size();
+  check_errors(total_dims, flip_dims_size, dims);
+
+  auto indices = std::vector<at::Tensor>(flip_dims_size);
+  for (int64_t i = 0; i < flip_dims_size; i++) {
+    indices[i] = at::arange(self.type().toScalarType(at::ScalarType::Long), self.size(i) - 1, -1, -1);
+  }
+  // creates a meshgrid
+  for (int64_t i = 0; i < flip_dims_size; i++) {
+    auto temp = std::vector<int64_t>(flip_dims_size, 1);
+    temp[i] = indices[i].size(0);
+    indices[i] = indices[i].view(IntList(temp));
+  }
+  return self.index(TensorList(indices));
```

Comment:
note that you need to permute the dimensions if they are not consecutive.
So for example, if you have
```python
a = torch.arange(2 * 3 * 4).reshape(2, 3, 4)
b = a.flip(0, 2)
```
this implementation will need to transpose the returned results because the advanced indexing puts all non-consecutive indices in the beginning of the tensor.
```python
b.shape # should be 2, 4, 3
result = b.permute(0, 2, 1)  # this is missing from the code!
```

Reply:
I am still missing something. In your code, how does `final_indices = [slice(i) for i in tensor.shape]` translate to ATen code? 

```python
def multi_meshgrid(*args):
    """
    Creates a meshgrid from possibly many
    elements (instead of only 2).
    Returns a nd tensor with as many dimensions
    as there are arguments
    """
    args = list(args)
    template = [1 for _ in args]
    for i in range(len(args)):
        n = args[i].shape[0]
        template_copy = template.copy()
        template_copy[i] = n
        args[i] = args[i].view(*template_copy)
        # there will be some broadcast magic going on
    return tuple(args)

def flip(tensor, dims):
    if not isinstance(dims, (tuple, list)):
        dims = [dims]
    indices = [torch.arange(tensor.shape[dim] - 1, -1, -1,
        dtype=torch.int64) for dim in dims]
    multi_indices = multi_meshgrid(*indices)
    final_indices = [slice(i) for i in tensor.shape]
    for i, dim in enumerate(dims):
        final_indices[dim] = multi_indices[i]
    flipped = tensor[final_indices]
    # need to permute the final dimensions
    # if dims is not consecutive, but I'm lazy
    # now :-)
    return flipped
```

Reply:
To get the final permutation order, all you need to do is to sort the indices after permutation.
say we select on dims 1 and 3, of a 5d tensor the end result of the flip will have the following order of dimensions
```python
flip_dims = [1, 3, 0, 2, 4, 5]
```
where `1, 3` comes first because of advanced indexing on non-consecutive dimensions, and then comes the other dimensions.
So this is obtained by something like
```python
dims + [i for i in range(tensor.dim()) if i not in dims]
```

Once you have this tensor containing the order of the dimensions, all you need to do is to get the indices after sorting:
```python
_, permutation = flip_dims.sort(0)
```

Note that those steps are only necessary if the dimensions in dim are non-consecutive, i.e., if
```python
torch.all(tensor(dims[1:] - dims[:-1]) < 2)
```
(or something like that).

Reply:
For the `slice(i)`, I believe all you need to do is to replace all the `slices` with undefined tensors, as that's what's used in the ATen backend to represent full slices I believe.

Reply:
@fmassa Thanks a lot! This works out perfectly now!